pubtools/pulplib/_impl/model/maintenance.py

```diff
@@ -0,0 +1,147 @@
+
+    last_updated_by = attr.ib(default="Content Delivery", type=str)
+    """Person/party who updated the report last time"""
+
+    repos = attr.ib(default={}, type=dict)
+    """A dictionary contains repo_id: :class:`MaintainedRepo` pairs, indicates
+    which repositories are in maintenance mode and details.
+    If empty, then it means no repositories is in maintenance mode.
+    """
+
+    @classmethod
+    def from_data(cls, data):
+        """Create a new report with a dictionary"""
+        try:
+            jsonschema.validate(instance=data, schema=cls._SCHEMA)
+        except jsonschema.exceptions.ValidationError as error:
+            LOG.exception("%s.from_data invoked with invalid data", cls.__name__)
+            raise InvalidDataException(str(error))
+
+        maintenance = cls(
+            last_updated=data["last_updated"],
+            last_updated_by=data["last_updated_by"],
+            repos={},
+        )
+        for repo_id, details in data["repos"].items():
+            maintenance.repos[repo_id] = MaintainedRepo(**details)
+
+        return maintenance
+
+    def json(self):
+        """Output a json format report.
+
+        Returns:
+            A string contains the maintenance report in json format.
+        """
+        report = {
+            "last_updated": self.last_updated,
+            "last_updated_by": self.last_updated_by,
+            "repos": {},
+        }
+
+        for repo_id, details in self.repos.items():
+            report["repos"].update(
+                {
+                    repo_id: {
+                        "message": details.message,
+                        "owner": details.owner,
+                        "started": details.started,
+                    }
+                }
+            )
+
+        return json.dumps(report, indent=4, sort_keys=True)
+
+    def add(self, repo_ids, **kwargs):
+        """Add entries to maintenance report and update the timestamp. Every
+        entry added to the report represents a repository in maintenace mode.
+
+        Args:
+            repo_ids (list[str]):
+            A list of repositories ids that will be added to the report.
+
+            Optional keyword args:
+                message (str):
+                    Reason why put the repo to maintenance.
+                owner (str):
+                    Who set the maintenance mode.
+
+        """
+        message = kwargs.get("message") or "Maintenance mode is enabled"
+        owner = kwargs.get("owner") or "Content Delivery"
+
+        for repo in repo_ids:
+            self.repos[repo] = MaintainedRepo(
+                owner=owner, message=message, started=_iso_time_now()
+            )
+
+        self._update_owner_time(owner)
+
+    def remove(self, regex, **kwargs):
+        """Remove entries from the maintenece report. Once the entry is removed,
+        it means the corresponding repository isn't in maintenance mode anymore.
+
+        Args:
+            regex (str):
```

Comment:
Why regex again?  It has the same problem I mentioned before that it's actually quite awkward to use, if you know exactly which repos you want to remove.